src/Features/Core/Portable/Completion/Providers/Snippets/AbstractSnippetProvider.cs

```diff
@@ -0,0 +1,97 @@
+{
+    internal abstract class AbstractSnippetProvider : ISnippetProvider
+    {
+        protected SyntaxAnnotation _cursorAnnotation = new();
+        protected SyntaxAnnotation _reformatAnnotation = new();
+
+        protected abstract Task<bool> IsValidSnippetLocationAsync(Document document, int position, CancellationToken cancellationToken);
+        protected abstract string GetSnippetDisplayName();
+        protected abstract Task<TextChange> GenerateSnippetTextChangeAsync(Document document, TextSpan span, int tokenSpanStart, int tokenSpanEnd, CancellationToken cancellationToken);
+        protected abstract Task<SyntaxNode> AnnotateRootForCursorAsync(Document document, TextSpan span, SyntaxAnnotation cursorAnnotation, CancellationToken cancellationToken);
+        protected abstract Task<SyntaxNode> AnnotateRootForReformattingAsync(Document document, TextSpan span, SyntaxAnnotation reformatAnnotation, CancellationToken cancellationToken);
+        protected abstract int GetTargetCaretPosition(SyntaxNode caretTarget);
+        protected abstract Task<ImmutableArray<TextSpan>> GetRenameLocationsAsync(Document document, TextSpan span, CancellationToken cancellationToken);
+
+        public async Task<SnippetData?> GetSnippetDataAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            if (await IsValidSnippetLocationAsync(document, position, cancellationToken).ConfigureAwait(false))
+            {
+                return new SnippetData(GetSnippetDisplayName());
+            }
+
+            return null;
+        }
+
+        public async Task<Snippet> GetSnippetAsync(Document document, TextSpan span, int tokenSpanStart, int tokenSpanEnd, CancellationToken cancellationToken)
+        {
+            var textChange = await GenerateSnippetTextChangeAsync(document, span, tokenSpanStart, tokenSpanEnd, cancellationToken).ConfigureAwait(false);
+            var snippetDocument = await GetDocumentWithSnippetAsync(document, span, textChange, tokenSpanStart, tokenSpanEnd, cancellationToken).ConfigureAwait(false);
+            var reformattedDocument = await ReformatDocumentAsync(snippetDocument, span, cancellationToken).ConfigureAwait(false);
+            var changes = await reformattedDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+            var changesArray = changes.ToImmutableArray();
+            var newText = await reformattedDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var change = Utilities.Collapse(newText, changesArray);
+            var annotatedRootForCursor = await AnnotateRootForCursorAsync(reformattedDocument, span, _cursorAnnotation, cancellationToken).ConfigureAwait(false);
+            reformattedDocument = reformattedDocument.WithSyntaxRoot(annotatedRootForCursor);
+            var caretTarget = annotatedRootForCursor.GetAnnotatedNodes(_cursorAnnotation).FirstOrDefault();
```

Comment:
Can we use `SingleOrDefault` here to help set expectations that only one node should be annotated for the cursor position?